### DISCUSSION TITLE
AMP support

### DISCUSSION THREAD

**Discussion by thilllon** [remix-run/remix#5225] [Q&A] [answered]
Hi, 
Our web service built with Remix needs to support Google AMP(https://developers.google.com/amp). We could not find AMP in the docs. It seems Remix does not support it for now. 
Is there any good idea to create AMP at once, not page-by-page? Any suggestion will be very helpful. Thanks

**Comment by sergiodxa** [comment, marked as answer]
Isn't AMP deprecated?

Aside of that, AMP is a web component library at its core, if you follow the [starter code guide on their docs](https://amp.dev/documentation/guides-and-tutorials/start/create/basic_markup), you can see that the only thing you need to do is to do a few changes in the HTML and load a script and inline style tag.

There's nothing in Remix that should prevent you to do that and then use the custom elements the library provides you.

**Reply by sergiodxa** [reply]
I just tried it and yes it's possible, I wrote about this here https://sergiodxa.com/articles/use-amp-with-remix

**Reply by huyb1991** [reply]
@sergiodxa Your article is just a sample, this won't work in real AMP page because it **not apply styles the same** to default website version, that is just HTML tag.

**Reply by huyb1991** [reply]
Should we re-open this discussion? IMO, AMP is popular for web at this time, many benefits for SEO, so this is important feature.

**Comment by wenzf** [comment]
@sergiodxa 

When doing:
`<style amp-boilerplate  />`
I get an error message: 

_"Warning: Received `true` for a non-boolean attribute `amp-boilerplate`.
If you want to write it to the DOM, pass a string instead: amp-boilerplate="true" or amp-boilerplate={value.toString()}."_

When doing `<style amp-boilerplate='' />` or `<style amp-boilerplate="true" />`, the AMP markup validator says that markup is not correct.

Note: When inspecting the page in the browser, `<style amp-boilerplate='' />` is shown as `<style amp-boilerplate />` => to reporoduce my issue, you need to look at the html output directly instead of inspecting it in the browser.

**Reply by sergiodxa** [reply]
That’s expected, AMP uses invalid HTML, they should have prefixed with data- every custom attribute.

**Reply by wenzf** [reply]
Is there a way to override this behaviour?
As it is now, it's not valid AMP and will not be indexed as such - at least according to the validator result. This would kll most if not all benefits of implementing AMP.

And thx for your work btw!

**Reply by sergiodxa** [reply]
Use just `amp-boilerplate` and accept the JS error, you shouldn't be shipping client-side React anyway in AMP so your HTML will remain exactly how it was generated on SSR.

But I'm not 100% sure since I don't use AMP myself, just tried the basics to see if it works, when I did I run the HTML through a validator and it was valid.

I would recommend you to don't use AMP, [Google is ditching it anyway](https://developers.google.com/search/blog/2021/04/more-details-page-experience) since it's not longer required for anything.

**Reply by wenzf** [reply]
> Use just `amp-boilerplate` and accept the JS error, you shouldn't be shipping client-side React anyway in AMP so your HTML will remain exactly how it was generated on SSR.

When doing this, only `<style />` without the `amp-boilerplate`  is rendered server side. - At least in my set up it.
 
> But I'm not 100% sure since I don't use AMP myself, just tried the basics to see if it works, when I did I run the HTML through a validator and it was valid.

Aside of that detail, everything seems to work fine

> I would recommend you to don't use AMP, [Google is ditching it anyway](https://developers.google.com/search/blog/2021/04/more-details-page-experience) since it's not longer required for anything.

I'm a aware of that. The idea is to additionally serve AMP versions of certain pages (news articles) to increase chances of getting indexed in Google News. In my understanding, it doesn't hurt in the worst case.

**Comment by huyb1991** [comment]
If we have a buffer layer/component, we can handle it, modify root.tsx is not allow us to do it since this apply for all and we have to choose one version for AMP or non-AMP version for all pages.

My idea is instead of `root.tsx` render HTML wrapper as current, we can pass it down one more lever (maybe in routers?), then we can create an `amp.tsx` router to container AMP tag, other AMP are happy with that, so, the routers structure can be:

```
- _index.tsx
- product.tsx
- product.$id.tsx
- amp.tsx
- amp.product.$id.tsx
```

**Reply by wenzf** [reply]
Yes, conditionally rendering pages as AMP or regular HTML doesn't work. And doing all pages as AMP is not an option.

I'm not sure if I can follow your idea. You think about rendering AMP and HTML separately and pass it down to one root.tsx? This way the root would have to deal with two separate versions as well. Or having two roots? 

My guess is that this would just lead to other issues. Maybe, doing a separate app only for AMP would be more pragmatic solution :D 

Let me know in case you find a working solution!

**Reply by huyb1991** [reply]
I think 2 roots can handle it but it make no sense for structure. I think some how we can override HTML structure and head is the good way, similar to override Meta component, so, in `routers/amp.tsx` file, we can override HTML by adding `amp` tag, and add header there, so other routers (should be child of `amp.tsx`) can extends this.

**Comment by RohitDeshwal** [comment]
I'm interested in incorporating Google AMP in my remix project. If anyone has recommendations or solutions to effectively implement AMP, I would greatly appreciate your guidance.

**Reply by wenzf** [reply]
Doing it like it's done here  [https://sergiodxa.com/articles/use-amp-with-remix](https://sergiodxa.com/articles/use-amp-with-remix) might work if the app is amp only. Good luck!
